src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1412,26 +1417,26 @@ private enum GetBytesMode { AllocateArray, Count, Span }
 
         /// <summary>
         /// Return the value of this BigInteger as a little-endian twos-complement
-        /// uint array, using the fewest number of uints possible. If the value is zero,
+        /// uint span, using the fewest number of uints possible. If the value is zero,
         /// return an array of one uint whose element is 0.
         /// </summary>
         /// <returns></returns>
-        private uint[] ToUInt32Array()
+        private ReadOnlySpan<uint> ToUInt32Array()
```

Comment:
Is the name with `Array`-suffix still a good choice?
As it's private, maybe it's possible to change the signature to something like
```c#
private bool TryToUInt32Span(Span<uint> buffer, out int written);
```
So a caller could provide a stack-alloced buffer, hence the allocation in L1458 could go away.

Reply:
This is the next part of optimization included in my implementation plan. Stack allocation is not enough here and I can use pooled arrays via `ArrayPool`. However I need advice about this approach.

Reply:
> need advice about this approach

The signature above seems fine for this.

The approach would be something like (somewhat generalized):

```c#
private const int MaxUintStackAllocCount = 256 / sizeof(uint);  // or any other reasonable constant

// ...

private bool TryToUInt32Span(Span<uint> buffer, out int written)
{
    // ...
}

public static BigInteger operator &(BigInteger left, BigInteger right)
{
    // ...
    int leftSize = /* determine the size needed */
    uint[] leftBufferFromPool = null;
    Span<uint> x = leftSize <= MaxUintStackAllocCount
        ? stackalloc uint[MaxUintStackAllocCount]       // stack-allocing a const is "better" that for a variable
        : leftBufferFromPool = ArrayPool<uint>.Shared.Rent(leftSize);

    // now we have a buffer for x that it as least leftSize large

    TryToUInt32Span(x, out int written);
    x = x.Slice(0, written);        // slice the buffer to the "actual" size, so further use is easier

    // similar for y and even z (as BigInteger has a ctor which takes a span)

    // use x, y, and z

    if (leftBufferFromPool != null)
    {
        ArrayPool<uint>.Shared.Return(leftBufferFromPool);
    }

    // similar for y and even z (as BigInteger has a ctor which takes a span)
}
```

If it's pretty common that the size is large enough, so stack-allocing doesn't make sense you can rent from the ArrayPool diretly.

Reply:
I'd suggest keeping this as returning a `uint[]` until it can be properly optimized (either via array pool or via taking in a `Span<T>`)

Reply:
So I see that pooled arrays are permitted. This is good news.
@gfoidl , `leftSize` in your example cannot be calculated prior to `TryToUInt32Span` call its value based on array from `left` after trimming procedure. So the only choice is to use size of the array even if it contains leading zeroes.

Reply:
The size just needs to be large enough to act as buffer and can hold all the data written to the Span. It can be sliced to the actual size afterwards, where a `written`-value (e.g. `out int written`) is set.

Thinks of it similar as in `Encoding.GetMaxBytes`. So if a bit larger than actual is fine :wink:
The ArrayPool will return a array sized to next power (= implementation detail, so could change) of two anyway, so this shouldn't be obstacle.